tensorboard/plugins/pr_curve/tf_pr_curve_dashboard/tf-pr-curve-dashboard.ts

```diff
@@ -0,0 +1,136 @@
+"use strict";
+import {RequestManager} from '../tf-backend/requestManager.js';
+import {getTags} from '../tf-backend/backend.js';
+import {getRouter} from '../tf-backend/router.js';
+import {categorizeTags} from '../tf-categorization-utils/categorizationUtils.js';
+import {runsColorScale} from "../tf-color-scale/colorScale.js";
+
+// Used to remember the step that each run is at. The runs may differ in
+// step for various reasons. One run might write summaries more often or
+// just run more slowly.
+interface RunStepPair {
+  run: string;
+  step: number;
+};
+
+Polymer({
+  is: 'tf-pr-curve-dashboard',
+  properties: {
+    _selectedRuns: Array,
+    _runToTag: Object,  // map<run: string, tags: string[]>
+    _dataNotFound: Boolean,
+    _tagFilter: {
+      type: String,  // upward bound from paper-input
+      value: '.*',
+    },
+    _categories: {
+      type: Array,
+      computed:
+        '_makeCategories(_runToTag, _selectedRuns, _tagFilter)',
+    },
+    _requestManager: {
+      type: Object,
+      value: () => new RequestManager(),
+    },
+    _step: {
+      type: Number,
+      value: 0,
+      notify: true,
+    },
+    // An array of steps (numbers) to show in the slider for the dashboard.
+    _stepsForSlider: {
+      type: Array,
+      notify: true,
+    },
+    _colorScale: {
+      type: Object,
+      value: () => ({scale: runsColorScale}),
+      readOnly: true,
+    },
+    _runToStepMapping: {
+      type: Object,
+      value: {},
+    },
+    _runStepPairs: {
+      type: Array,
+      notify: true,
+    },
+  },
+  listeners: {
+    'data-updated': '_handleNewRunToDataMapping',
```

Comment:
As discussed offline, please remove these listeners and make all dataflow downward-only. This will significantly simplify the process of following the data flow, which has been one of my main goals for the past few months.

I will review the rest of this component once you make this change. It's too hard for me to follow the data flow as it is.

One general comment: the data flow in this component is also highly imperative, and would be further easier to understand if it were more declarative. Instead of functions called `updateFoo`, I want to see `computeFoo`. The change to downward-only flow will make this easier to implement.